commands/template_store_list.go

```diff
@@ -0,0 +1,170 @@
+	officialRepo        = "openfaas"
+	allTemplates        = "all"
+)
+
+var (
+	source string
+)
+
+func init() {
+	templateStoreListCmd.Flags().StringVarP(&source, "source", "s", allTemplates, "The source of the templates")
+	templateStoreListCmd.Flags().BoolVarP(&verbose, "verbose", "v", false, "Shows additional language and platform")
+
+	templateStoreCmd.AddCommand(templateStoreListCmd)
+}
+
+var templateStoreListCmd = &cobra.Command{
+	Use:     `list`,
+	Short:   `List templates from OpenFaaS organizations`,
+	Aliases: []string{"ls"},
+	Long:    `List templates supported by openfaas or openfaas-incubator organizations or both`,
+	Example: `  faas-cli template store list --source=openfaas --verbose
+  faas-cli template store ls --source=openfaas-incubator -v
+  faas-cli template store ls --source=all`,
+	RunE: runTemplateStoreList,
+}
+
+func runTemplateStoreList(cmd *cobra.Command, args []string) error {
+	templatesInfo, templatesErr := getTemplateInfo(DefaultTemplateList)
+	if templatesErr != nil {
+		return fmt.Errorf("error while getting templates info: %s\n", templatesErr)
+	}
+
+	formattedOutput := formatTemplatesOutput(templatesInfo)
+
+	fmt.Println(formattedOutput)
+	return nil
+}
+
+func getTemplateInfo(repository string) (TemplatesInfo, error) {
+	req, reqErr := http.NewRequest(http.MethodGet, repository, nil)
+	if reqErr != nil {
+		return nil, fmt.Errorf("error while trying to create request to take template info: %s\n", reqErr.Error())
+	}
+
+	client := http.DefaultClient
+	res, clientErr := client.Do(req)
+	if clientErr != nil {
+		return nil, fmt.Errorf("error while requesting template list: %s\n", clientErr.Error())
+	}
+
+	if res != nil {
+		defer res.Body.Close()
+		if res.StatusCode != http.StatusOK {
+			return nil, fmt.Errorf("unexpected status code wanted: %d got: %d\n", http.StatusOK, res.StatusCode)
+		}
+
+		body, bodyErr := ioutil.ReadAll(res.Body)
+		if bodyErr != nil {
+			return nil, fmt.Errorf("error while reading data from templates body: %s\n", bodyErr.Error())
+		}
+
+		templatesInfo := TemplatesInfo{}
+		unmarshallErr := json.Unmarshal(body, &templatesInfo)
+		if unmarshallErr != nil {
+			return nil, fmt.Errorf("error while unmarshalling into templates struct: %s\n", unmarshallErr.Error())
+		}
+		return templatesInfo, nil
+	}
+	return nil, fmt.Errorf("error empty response body from: %s\n", DefaultTemplateList)
+}
+
+func formatTemplatesOutput(templates TemplatesInfo) string {
+	if source != allTemplates && source != incubatorRepo && source != officialRepo {
+		fmt.Printf("\nThe --source flag needs to have value %s, %s or %s\n", allTemplates, incubatorRepo, officialRepo)
+		return ""
+	}
+	var buff bytes.Buffer
+	lineWriter := tabwriter.NewWriter(&buff, 0, 0, 1, ' ', 0)
+	fmt.Fprintln(lineWriter)
+	if verbose {
+		fmt.Fprintf(lineWriter, "NAME\tLANGUAGE\tPLATFORM\tSOURCE\tDESCRIPTION\n")
+	} else {
+		fmt.Fprintf(lineWriter, "NAME\tSOURCE\tDESCRIPTION\n")
+	}
+	for _, template := range templates {
```

Comment:
I think we could reduce the number of blocks/levels here by breaking this into a couple steps (1) filter the list, then (2) print the list.  This results in two loops.  Alternatively or at the same time, we could also move some of these blocks to new methods. This will also improve the readability and some repetition.  This loop could look something like 

```golang
for _, template := range templates {
    if skipSource(template, source) {
        continue
    }
    printTemplate(writer, template, verbose)
}
```

Reply:
Great Idea, thanks for the suggestion

Reply:
Also done ^ moved the parts into separate methods (they still have repetition, I cannot escape that with the source ...)